stdlib/public/Concurrency/TaskStatus.cpp

```diff
@@ -576,58 +550,79 @@ static void performEscalationAction(TaskStatusRecord *record,
-  // as high as the desired priority.
-  auto status = task->_private().Status.load(std::memory_order_relaxed);
-  if (status.getStoredPriority() >= newPriority)
-    return status.getStoredPriority();
-
-  withStatusRecordLock(task, LockContext::OtherAsynchronous, status, [&] {
-    // Now that we have the task's status lock, check again that the
-    // priority is still too low.
-    if (status.getStoredPriority() >= newPriority)
-      return;
-    status = status.withEscalatedPriority(newPriority);
 
-    // TODO: attempt to escalate the thread running the task, if it's
-    // currently running.  This probably requires the task to be enqueued
-    // on a standard executor.
+  auto oldStatus = task->_private().Status.load(std::memory_order_relaxed);
+  auto newStatus = oldStatus;
 
-    // Perform escalation operations for all the status records.
-    for (auto cur: status.records()) {
-      performEscalationAction(cur, newPriority);
+  while (true) {
+    // Fast path: check that the stored priority is already at least
+    // as high as the desired priority.
+    if (oldStatus.getStoredPriority() >= newPriority)
+      return oldStatus.getStoredPriority();
+
+    // Regardless of whether status record is locked or not, update the priority
+    // and RO bit on the task status
+    if (oldStatus.isRunning()) {
+      newStatus = oldStatus.withEscalatedPriority(newPriority);
+    } else {
+      newStatus = oldStatus.withNewPriority(newPriority);
     }
-  });
-
-  return status.getStoredPriority();
-}
-
-void AsyncTask::flagAsRunning_slow() {
-  withStatusRecordLock(this, LockContext::OnTask,
-                       [&](ActiveTaskStatus &status) {
-    assert(!status.isRunning());
 
-    status = status.withRunning(true);
-    if (status.isStoredPriorityEscalated()) {
-      status = status.withoutStoredPriorityEscalation();
-      Flags.setPriority(status.getStoredPriority());
-      concurrency::trace::task_flags_changed(this, Flags.getOpaqueValue());
+    if (task->_private().Status.compare_exchange_weak(oldStatus, newStatus,
+            /* success */ std::memory_order_relaxed,
+            /* failure */ std::memory_order_relaxed)) {
+      break;
     }
-  });
-}
-
-void AsyncTask::flagAsSuspended_slow() {
-  withStatusRecordLock(this, LockContext::OnTask,
-                       [&](ActiveTaskStatus &status) {
-    assert(status.isRunning());
+  }
 
-    status = status.withRunning(false);
-    if (status.isStoredPriorityEscalated()) {
-      status = status.withoutStoredPriorityEscalation();
-      Flags.setPriority(status.getStoredPriority());
-      concurrency::trace::task_flags_changed(this, Flags.getOpaqueValue());
+#if SWIFT_CONCURRENCY_ENABLE_PRIORITY_ESCALATION
+  if (newStatus.isRunning()) {
+    // The task is running, escalate the thread that is running it. The
+    // dispatch SPI does an atomic debounce on the lock address to make sure
+    // that the thread that is expected to be the drainer is still the
```

Comment:
Trailing sentence

Is "debounce" Dispatch jargon for this?  As a general concept, debouncing means combining multiple events into one, which is not what this is about — this is about making sure that the thread priority isn't changed for stale reasons.

Reply:
Hmmm, let me check with folks offline.

Reply:
Looking at the linux futex docs (https://man7.org/linux/man-pages/man2/futex.2.html), they don't seem to have a singular word or phrase that really describes what the operation is trying to do: 

```
When executing a futex operation that requests to block a thread,
the kernel will block only if the futex word has the value that
the calling thread supplied (as one of the arguments of the
futex() call) as the expected value of the futex word.  The
loading of the futex word's value, the comparison of that value
with the expected value, and the actual blocking will happen
atomically and will be totally ordered with respect to concurrent
operations performed by other threads on the same futex word.
Thus, the futex word is used to connect the synchronization in
user space with the implementation of blocking by the kernel.
Analogously to an atomic compare-and-exchange operation that
potentially changes shared memory, blocking via a futex is an
atomic compare-and-block operation.
```

@rjmccall I'm open to suggestions of how else to succintly name the operation described above.

Alternatively, I'm inclined to just remove this comment altogether since it's really about documenting what dispatch is doing now and might want to do in the future. I put it here cause I was co-developing the dispatch interface at the same time and so the lines were blurry to me.